grants_tagger/tune_threshold.py

```diff
@@ -0,0 +1,134 @@
+"""
+Tune threshold of multilabel classifier to maximise f1
+
+based on paper "Threshold optimisation for multi-label classifiers"
+by Pillai https://doi.org/10.1016/j.patcog.2013.01.012
+
+There are currently two deviations from the paper
+* fixed number of thresholds to be tried instead of all values from Y_pred_proba for efficiency
+* initialisation of thresholds with 0.5 which seems to help convergence in large number of labels
+"""
+from pathlib import Path
+import argparse
+import pickle
+import random
+
+from sklearn.model_selection import train_test_split
+from sklearn.metrics import f1_score, multilabel_confusion_matrix, confusion_matrix
+from scipy.sparse import csr_matrix, issparse
+import numpy as np
+
+from grants_tagger.evaluate_model import predict
+from grants_tagger.utils import load_data
+
+
+def argmaxf1(thresholds, Y_test, Y_pred_proba, label_i, nb_thresholds, iterations):
+    candidate_thresholds = thresholds.copy()
+    current_threshold_i = thresholds[label_i]
+    max_f1 = 0
+    optimal_threshold_i = current_threshold_i
+
+    Y_pred = Y_pred_proba > candidate_thresholds
+    Y_pred = csr_matrix(Y_pred)
+    cm = multilabel_confusion_matrix(Y_test, Y_pred)
+
+    if nb_thresholds:
+        candidate_thresholds_i = [t/nb_thresholds for t in range(1, nb_thresholds)]
+    else:
+        candidate_thresholds_i = Y_pred_proba[:, label_i]
```

Comment:
Might be worth `np.unique(Y_pred_proba[:, label_i])`?

Reply:
good idea

Reply:
8b1752b